lib/note-detail.jsx

```diff
@@ -1,62 +1,166 @@
-import React, { PropTypes } from 'react';
-import marked from 'marked';
+import React, { Component, PropTypes } from 'react';
 import Textarea from 'react-textarea-autosize';
-import { noop, get, debounce } from 'lodash';
 import analytics from './analytics';
+import marked from 'marked';
+import { get, debounce, invoke } from 'lodash';
 import { viewExternalUrl } from './utils/url-utils';
 
-const uninitializedNoteEditor = { focus: noop };
 const saveDelay = 2000;
 
-export default React.createClass( {
+const isValidNote = note => note && note.id;
+const prependTab = l => `\t${ l }`;
+const removeLeadingTab = l => l && l.length && l[0] === '\t' ? l.substring( 1 ) : l;
 
-	propTypes: {
+export default class NoteDetail extends Component {
+	static propTypes = {
 		note: PropTypes.object,
 		previewingMarkdown: PropTypes.bool,
 		fontSize: PropTypes.number,
 		onChangeContent: PropTypes.func.isRequired
-	},
+	};
+
+	constructor( props ) {
+		super( props );
 
-	componentWillMount: function() {
+		this.storeNoteEditor = r => this.noteEditor = r;
 		this.queueNoteSave = debounce( this.saveNote, saveDelay );
-		this.noteEditor = uninitializedNoteEditor;
-	},
+	}
 
-	componentDidMount: function() {
+	componentDidMount = () => {
 		// Ensures note gets saved if user abruptly quits the app
 		window.addEventListener( 'beforeunload', this.queueNoteSave.flush );
-	},
-
-	initializeNoteEditor: function( noteEditor ) {
-		this.noteEditor = noteEditor;
-	},
-
-	isValidNote: function( note ) {
-		return note && note.id;
-	},
+		window.addEventListener( 'keydown', this.interceptTabPresses );
+	};
 
-	componentWillReceiveProps: function() {
+	componentWillReceiveProps = () => {
 		this.queueNoteSave.flush();
-	},
+	};
 
-	componentDidUpdate: function() {
+	componentDidUpdate = () => {
 		const { note } = this.props;
 		const content = get( note, 'data.content', '' );
-		if ( this.isValidNote( note ) && this.noteEditor ) {
+		if ( isValidNote( note ) && this.noteEditor ) {
 			this.noteEditor.value = content;
 
 			// Let's focus the editor for new and empty notes
 			if ( content === '' ) {
-				this.noteEditor.focus();
+				invoke( this.noteEditor, 'focus' );
 			}
 		}
-	},
+	};
 
-	componentWillUnmount: function() {
+	componentWillUnmount = () => {
 		window.removeEventListener( 'beforeunload', this.queueNoteSave.flush );
-	},
+		window.removeEventListener( 'keydown', this.interceptTabPresses );
+	};
+
+	handleKey = event => {
+		if (
+			'Enter' !== event.key ||
+			! this.noteEditor ||
+			this.props.previewingMarkdown ||
+			event.shiftKey ||
+			event.altKey ||
+			event.ctrlKey ||
+			event.metaKey
+		) {
+			return true;
+		}
+
+		const {
+			selectionStart,
+			value,
+		} = this.noteEditor;
+
+		const prevLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+		const prevLine = value.substring( prevLineStart, selectionStart - 1 );
+		const prevIndent = prevLine.match( /^(\s+)([-*])/ );
```

Comment:
@nfcampos did you find a problem with it? inside of the character list `[]` some of the RegExp symbols take on new meaning, some are even position-dependent. In this case `*` matches a literal `*` and _doesn't mean_ "any number of `-`"